server/src/state.rs

```diff
@@ -70,8 +77,38 @@ impl Client {
 
                         let mut sockets = sockets.lock().await;
                         let sockets_len = sockets.len();
+                        if sockets_len >= max_sockets as usize {
+                            // reached sockets max but check if there are timeout connections we can remove to make room
+                            let mut connected_sockets = vec![];
+                            while let Some(s) = sockets.pop() {
```

Comment:
What do you mean? Which callback?

Reply:
> I actually don't know if it's possible to register a function to passively clean the closed socket.

I don't think so, tokio is not a callback oriented API. And in any case we need to drop our socket object to clean it.

> Just think that iterate and find the closed sockets seems a bit risky and may result into DoS.

How so? Why would this be more vulnerable to DoS than the existing code?

Reply:
A client can already spam connections and starve the connections pool, making the tunnel unusable. I don't think this patch is making DoS any worst tbh.

> We need a server controlled method to clean the dead sockets.

The only option I can think of is a timer regularly checking and purging dead sockets.

Reply:
I introduced this patch to fix the following case I hit during my tests:

- client establish tunnel with, say, 10 connections on a max of 10 allowed
- disconnect the client from the network until the TCP keepalive timesout
- `copy_bidirectional()` fails and so client tries to re-establish those connections
- When client is back on the network these new connections are rejected by the server because the 10 allowed connections are still filled by the dead ones and so the tunnel is unusable.

Indeed if the client tries to actually use the tunnel, `take`  will cleanup those connections but that means this first attempt to use the tunnel will fail and the tunnel will remain unusable until the next attempt to re-establish the connections.

So I think we do need to be as proactive as possible when cleaning up dead connections to minimize the tunnel down time in such  scenario.

Are you ok if I introduce such timer then instead of the code in the last patch?

Reply:
I replaced my last commit which is now using a timer.